In [2]:
import pandas as pd


In [4]:
df = pd.read_csv("Auto_mobile.csv")
df.head()

Unnamed: 0  symboling  normalized-losses         make aspiration  \
0           0          3                122  alfa-romero        std   
1           1          3                122  alfa-romero        std   
2           2          1                122  alfa-romero        std   
3           3          2                164         audi        std   
4           4          2                164         audi        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   

   city-mpg  highway-mpg    price  highway-L/100Km city-L/100km  \
0        21           27  13495.0         8.703704    11.190476   
1        21           27  16500.0         8.703704    11.190476   
2        19           26  16500.0         9.038462    12.368421   
3        24           30  13950.0         7.833333     9.791667   
4        18           22  17450.0        10.681818    13.055556   

  horsepower-binned  fuel-type-diesel fuel-type-gas  aspiration-std  \
0               Low                 0             1               1   
1               Low                 0             1               1   
2            Medium                 0             1               1   
3               Low                 0             1               1   
4               Low                 0             1               1   

   aspiration-turbo  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  

[5 rows x 33 columns]

<b><h2 id="discriptive_statistics">Descriptive Statistical Analysis</h2></b>

<p>Let's first take a look at the variables by utilizing a description method.</p>

<p>The <b>describe</b> function automatically computes basic statistics for all continuous variables. Any NaN values are automatically skipped in these statistics.</p>

This will show:
<ul>
    <li>the count of that variable</li>
    <li>the mean</li>
    <li>the standard deviation (std)  $SD=\sqrt{\frac{\sum_{i=1}^N({x_i-\bar{x})^2}}{N-1}}$  </li> 
    <li>the minimum value</li>
    <li>the IQR (Interquartile Range: 25%, 50% and 75%)</li>
    <li>the maximum value</li>
<ul>
 We can apply the method "describe" as follows:

In [6]:
df.describe()

Unnamed: 0   symboling  normalized-losses  wheel-base      length  \
count  201.000000  201.000000          201.00000  201.000000  201.000000   
mean   100.000000    0.840796          122.00000   98.797015    0.837102   
std     58.167861    1.254802           31.99625    6.066366    0.059213   
min      0.000000   -2.000000           65.00000   86.600000    0.678039   
25%     50.000000    0.000000          101.00000   94.500000    0.801538   
50%    100.000000    1.000000          122.00000   97.000000    0.832292   
75%    150.000000    2.000000          137.00000  102.400000    0.881788   
max    200.000000    3.000000          256.00000  120.900000    1.000000   

            width      height  curb-weight  engine-size        bore  ...  \
count  201.000000  201.000000   201.000000   201.000000  201.000000  ...   
mean     0.915126    0.899108  2555.666667   126.875622    3.330692  ...   
std      0.029187    0.040933   517.296727    41.546834    0.268072  ...   
min      0.837500    0.799331  1488.000000    61.000000    2.540000  ...   
25%      0.890278    0.869565  2169.000000    98.000000    3.150000  ...   
50%      0.909722    0.904682  2414.000000   120.000000    3.310000  ...   
75%      0.925000    0.928094  2926.000000   141.000000    3.580000  ...   
max      1.000000    1.000000  4066.000000   326.000000    3.940000  ...   

          peak-rpm    city-mpg  highway-mpg         price  highway-L/100Km  \
count   201.000000  201.000000   201.000000    201.000000       201.000000   
mean   5117.665368   25.179104    30.686567  13207.129353         8.044957   
std     478.113805    6.423220     6.815150   7947.066342         1.840739   
min    4150.000000   13.000000    16.000000   5118.000000         4.351852   
25%    4800.000000   19.000000    25.000000   7775.000000         6.911765   
50%    5125.369458   24.000000    30.000000  10295.000000         7.833333   
75%    5500.000000   30.000000    34.000000  16500.000000         9.400000   
max    6600.000000   49.000000    54.000000  45400.000000        14.687500   

       city-L/100km  fuel-type-diesel  fuel-type-gas  aspiration-std  \
count    201.000000        201.000000     201.000000      201.000000   
mean       9.944145          0.099502       0.900498        0.820896   
std        2.534599          0.300083       0.300083        0.384397   
min        4.795918          0.000000       0.000000        0.000000   
25%        7.833333          0.000000       1.000000        1.000000   
50%        9.791667          0.000000       1.000000        1.000000   
75%       12.368421          0.000000       1.000000        1.000000   
max       18.076923          1.000000       1.000000        1.000000   

       aspiration-turbo  
count        201.000000  
mean           0.179104  
std            0.384397  
min            0.000000  
25%            0.000000  
50%            0.000000  
75%            0.000000  
max            1.000000  

[8 rows x 23 columns]

In [7]:
df.describe(include=['object'])

make aspiration num-of-doors body-style drive-wheels  \
count      201        201          201        201          201   
unique      22          2            2          5            3   
top     toyota        std         four      sedan          fwd   
freq        32        165          115         94          118   

       engine-location engine-type num-of-cylinders fuel-system  \
count              201         201              201         201   
unique               2           6                7           8   
top              front         ohc             four        mpfi   
freq               198         145              157          92   

       horsepower-binned  
count                201  
unique                 3  
top                  Low  
freq                 153

In [8]:
df.describe(include='all')

Unnamed: 0   symboling  normalized-losses    make aspiration  \
count   201.000000  201.000000          201.00000     201        201   
unique         NaN         NaN                NaN      22          2   
top            NaN         NaN                NaN  toyota        std   
freq           NaN         NaN                NaN      32        165   
mean    100.000000    0.840796          122.00000     NaN        NaN   
std      58.167861    1.254802           31.99625     NaN        NaN   
min       0.000000   -2.000000           65.00000     NaN        NaN   
25%      50.000000    0.000000          101.00000     NaN        NaN   
50%     100.000000    1.000000          122.00000     NaN        NaN   
75%     150.000000    2.000000          137.00000     NaN        NaN   
max     200.000000    3.000000          256.00000     NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           201        201          201             201  201.000000  ...   
unique            2          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            115         94          118             198         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.797015  ...   
std             NaN        NaN          NaN             NaN    6.066366  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

          city-mpg  highway-mpg         price  highway-L/100Km city-L/100km  \
count   201.000000   201.000000    201.000000       201.000000   201.000000   
unique         NaN          NaN           NaN              NaN          NaN   
top            NaN          NaN           NaN              NaN          NaN   
freq           NaN          NaN           NaN              NaN          NaN   
mean     25.179104    30.686567  13207.129353         8.044957     9.944145   
std       6.423220     6.815150   7947.066342         1.840739     2.534599   
min      13.000000    16.000000   5118.000000         4.351852     4.795918   
25%      19.000000    25.000000   7775.000000         6.911765     7.833333   
50%      24.000000    30.000000  10295.000000         7.833333     9.791667   
75%      30.000000    34.000000  16500.000000         9.400000    12.368421   
max      49.000000    54.000000  45400.000000        14.687500    18.076923   

       horsepower-binned  fuel-type-diesel fuel-type-gas  aspiration-std  \
count                201        201.000000    201.000000      201.000000   
unique                 3               NaN           NaN             NaN   
top                  Low               NaN           NaN             NaN   
freq                 153               NaN           NaN             NaN   
mean                 NaN          0.099502      0.900498        0.820896   
std                  NaN          0.300083      0.300083        0.384397   
min                  NaN          0.000000      0.000000        0.000000   
25%                  NaN          0.000000      1.000000        1.000000   
50%                  NaN          0.000000      1.000000        1.000000   
75%                  NaN          0.000000      1.000000        1.000000   
max                  NaN          1.000000      1.000000        1.000000   

        aspiration-turbo  
count         201.000000  
unique               NaN  
top                  NaN  
freq                 NaN  
mean            0.179104  
std             0.384397  
min             0.000000  
25%             0.000000  
50%             0.000000  
75%             0.000000  
max             1.00000

<b><h5>Value Counts</h5></b>
<p>Value-counts is a good way of understanding how many units of each characteristic/variable we have. We can apply the "value_counts" method on the column 'engine-location'. Don’t forget the method "value_counts" only works on Pandas series, not Pandas Dataframes. As a result, we only include one bracket "df['engine-location']" not two brackets "df[['engine-location']]".</p>

In [9]:
df['engine-location'].value_counts()

front    198
rear       3
Name: engine-location, dtype: int64

We can convert the series to a Dataframe and save the results to the dataframe "engine_loc_counts" and rename the column  'engine-location' to 'value_counts'. Then let's rename the index to 'engine-location':

In [10]:
engine_loc_counts = df['engine-location'].value_counts().to_frame()
engine_loc_counts

engine-location
front              198
rear                 3

In [11]:
engine_loc_counts.rename(columns={'engine-location': 'value_counts'}, inplace=True)
engine_loc_counts

value_counts
front           198
rear              3

In [12]:
engine_loc_counts.index.name = 'engine-location'
engine_loc_counts.head()

value_counts
engine-location              
front                     198
rear                        3

<p>Examining the value counts of the engine location would not be a good predictor variable for the price. This is because we only have three cars with a rear engine and 198 with an engine in the front, this result is skewed. Thus, we are not able to draw any conclusions about the engine location.</p>

<b><h2 id="basic_grouping">Basics of Grouping</h2></b>

<p>The "groupby" method groups data by different categories. The data is grouped based on one or several variables and analysis is performed on the individual groups.</p>

<p>For example, let's group by the variable "drive-wheels". We see that there are 3 different categories of drive wheels.</p>

In [13]:
df['drive-wheels'].unique()

array(['rwd', 'fwd', '4wd'], dtype=object)

Assume you want to know, is there any relationship between the different types of drive system, forward, rear, and four-wheel drive, and the price of the vehicles? If so, which type of drive system adds the most value to a vehicle? It would be nice if we could group all the data by the different types of drive wheels and compare the results of these different drive wheels against each other.we can group "drive-wheels" and then average them.
We can select the columns 'drive-wheels', 'body-style' and 'price', then assign it to the variable "df_group_one".

In [14]:
df_group_one = df[['drive-wheels','price']]

In [15]:
# grouping results
df_group_one = df_group_one.groupby(['drive-wheels'],as_index=False).mean()
df_group_one

drive-wheels         price
0          4wd  10241.000000
1          fwd   9244.779661
2          rwd  19757.613333

In [16]:
# grouping results
df_gptest = df[['drive-wheels','body-style','price']]
grouped_test1 = df_gptest.groupby(['drive-wheels','body-style'],as_index=False).mean()
grouped_test1

drive-wheels   body-style         price
0           4wd    hatchback   7603.000000
1           4wd        sedan  12647.333333
2           4wd        wagon   9095.750000
3           fwd  convertible  11595.000000
4           fwd      hardtop   8249.000000
5           fwd    hatchback   8396.387755
6           fwd        sedan   9811.800000
7           fwd        wagon   9997.333333
8           rwd  convertible  23949.600000
9           rwd      hardtop  24202.714286
10          rwd    hatchback  14337.777778
11          rwd        sedan  21711.833333
12          rwd        wagon  16994.222222

In [17]:
grouped_pivot = grouped_test1.pivot(index='drive-wheels',columns='body-style')
grouped_pivot

price                                            \
body-style   convertible       hardtop     hatchback         sedan   
drive-wheels                                                         
4wd                  NaN           NaN   7603.000000  12647.333333   
fwd              11595.0   8249.000000   8396.387755   9811.800000   
rwd              23949.6  24202.714286  14337.777778  21711.833333   

                            
body-style           wagon  
drive-wheels                
4wd            9095.750000  
fwd            9997.333333  
rwd           16994.222222

In [18]:
grouped_pivot = grouped_pivot.fillna(0) #fill missing values with 0
grouped_pivot

price                                            \
body-style   convertible       hardtop     hatchback         sedan   
drive-wheels                                                         
4wd                  0.0      0.000000   7603.000000  12647.333333   
fwd              11595.0   8249.000000   8396.387755   9811.800000   
rwd              23949.6  24202.714286  14337.777778  21711.833333   

                            
body-style           wagon  
drive-wheels                
4wd            9095.750000  
fwd            9997.333333  
rwd           16994.222222